EF6.PG/NpgsqlMigrationSqlGenerator.cs

```diff
@@ -551,6 +551,11 @@ void AppendColumn(ColumnModel column, StringBuilder sql)
                     break;
                 }
             }
+            else if (column.IsNullable != null && !column.IsNullable.Value && column.ClrDefaultValue != null)
```

Comment:
This simplification can also be done above:

```suggestion
            else if (column.IsNullable == false && column.ClrDefaultValue != null)
```

I wonder if anything is done for reference types, specifically string where an empty string can be used as the default. Any chance you can take a look at what the SQL Server provider is doing on this?

Reply:
> I wonder if anything is done for reference types, specifically string where an empty string can be used as the default. Any chance you can take a look at what the SQL Server provider is doing on this?

There is no simple way to specify the default value of table column at database level in EF6, so I don't think this can be an issue at all. 

The [DefaultValue("")] data annotation does not set property's column default value on database level on any provider, including SQL Server provider.

Reply:
I'm not sure I understand... If you make an nullable int column, this PR makes it have 0 as its default value, right? Doesn't it make sense to do the same for empty string?

Reply:
> I'm not sure I understand... If you make an nullable int column, this PR makes it have 0 as its default value, right? Doesn't it make sense to do the same for empty string?

Here's what happens when you add "int" column without any annotations:

```C#
            if (column.IsNullable != null && !column.IsNullable.Value) //This will be true
                sql.Append(" NOT NULL");

            if (column.DefaultValue != null) //this will be false
            {
                sql.Append(" DEFAULT ");
                AppendValue(column.DefaultValue, sql);
            }
            else if (!string.IsNullOrWhiteSpace(column.DefaultValueSql)) //this will be false
            {
                sql.Append(" DEFAULT ");
                sql.Append(column.DefaultValueSql);
            }
            else if (column.IsIdentity) //this will also be false
            {
                switch (column.Type)
                {
                case PrimitiveTypeKind.Guid:
                    //CreateExtension("uuid-ossp");
                    //If uuid-ossp is not enabled migrations throw exception
                    AddStatment("select * from uuid_generate_v4()");
                    sql.Append(" DEFAULT uuid_generate_v4()");
                    break;
                case PrimitiveTypeKind.Byte:
                case PrimitiveTypeKind.SByte:
                case PrimitiveTypeKind.Int16:
                case PrimitiveTypeKind.Int32:
                case PrimitiveTypeKind.Int64:
                    //TODO: Add support for setting "SERIAL"
                    break;
                }
            }
```

So we end up with "ALTER TABLE table ADD COLUMN column int4 NOT NULL, which results in error if there's already records in table, because the DB system does not know what value to set for this column on already existing records.

int? is nullable so there will not be a problem, because "NOT NULL" will not be added to the definition, and the system will set null for all present records. String is also nullable, so there will not be a problem either - same scenario.

The issue #106 only counts on non-nullable types, when there's already records in table. The resolution only adds Clr default value for non-nullable types if either "Column.DefaultValue" and "Column.DefaultValueSql" are not set. The SQL Server provider and MySQL provider are doing the same thing.